### PR TITLE
Inline Help: Add e2e tests for inline help admin section

### PR DESCRIPTION
**Pull request by jeryj**
When we added the [admin section search results](https://github.com/Automattic/wp-calypso/pull/43862), we didn't update the e2e tests to account for them. This PR checks for the admin search results section as well as adds more assertions over what should and not be present at different times.

 
#### Changes proposed in this Pull Request
- [ ] ~~The test for the "No results" message needs to assert on the text content not the CSS selector. This will be more consistent with what a user sees.~~ I'm relying on the CSS selector for now.
- [x] Assert on the presence of the two sections `Support articles` and `Show me where to`.
- [x] Assert on the presence of admin links in search results.
- [x] Assert on variations and combinations of support links and admins links being shown or not shown.
- [x] Improve selectors in the `SupportSearchComponent` to prefer selecting by `aria-role` or text instead of CSS selector. This will make them more in line with what a user perceives rather than testing the implementation details.
- [x] Changes the `aria-label` on the no-search state to be `Helpful resources for this section` instead of `Search Results`. It becomes `Search Results` only if a search has API results.

#### Testing instructions

* Boot Calypso from the project root `yarn && yarn start`.
* Set-up e2e testing for Calypso
* `cd test/e2e`
* `yarn`
* `./node_modules/.bin/mocha specs/wp-inline-help-support-search-spec.js` <-- All should pass
* `./node_modules/.bin/mocha specs/wp-my-home-support-search-spec.js`<--- All should pass

* Manual for the `aria-label`: Go to https://calypso.live/?branch=add/admin-search-e2e
* Click "My Home"
* Inspect the search card component "results" and see the `aria-label` for the results says `Helpful resources for this section`.
* Perform a search with results, like "Domain"
* The `aria-label` should now be "Search results"
* Search for something that will not return results, like "lkjakdfjasdf"
* The `aria-label` for the results should be `Helpful resources for this section` again.

Fixes https://github.com/Automattic/wp-calypso/issues/44169

### CODE REVIEW

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 1]
```diff
@@ -38,28 +35,9 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
-		assert.equal(
-			isToggleVisible,
-			true,
-			'Inline Help support search was not displayed on a non My Home page.'
-		);
-	} );
-
-	step( 'Select the My Home page', async function () {
-		const sidebarComponent = await SidebarComponent.Expect( driver );
-		await sidebarComponent.selectMyHome();
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:62]: Not sure why you're removing this here?

I really don't think you can remove this code without reinstating it a followup PR.

It's really important we validate somewhere that the FAB doesn't appear on My Home. Personally I still see this test as the logical home for this.

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 2]
```diff
@@ -78,46 +56,32 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
-
-		// Once removed we can assert is is invisible.
-		const isToggleVisible = await inlineHelpPopoverComponent.isToggleVisible();
-		assert.equal(
-			isToggleVisible,
-			false,
-			'Inline Help support search was not correctly hidden on Home page.'
-		);
-	} );
-
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:94]: Same here. We need this test. Not sure why it's removed.

File: test/e2e/specs/wp-my-home-support-search-spec.js (hunk 1 above)
Comment (jeryj) [reply test/e2e/specs/wp-my-home-support-search-spec.js:62]: I felt like it was more appropriate on the FAB, as the FAB and the search component don't really have anything to do with each other, IMO. If there were an issue with the FAB showing up on the `/home` route, it would be a FAB issue, not a support search issue, so I was planning to put it in the FAB related tests.

File: test/e2e/specs/wp-inline-help-support-search-spec.js [hunk 3]
```diff
@@ -28,49 +29,56 @@ before( async function () {
...
+			const isToggleVisible = await inlineHelpPopoverComponent.isToggleVisible();
+			assert.equal(
+				isToggleVisible,
+				true,
+				'Inline Help support search was not shown on Theme page.'
+			);
 		} );
 	} );
 
 	describe( 'Performing searches', function () {
```
Comment (jeryj) [inline test/e2e/specs/wp-inline-help-support-search-spec.js:74]: @getdave I feel like almost everything in this section could get removed since it's tested more thoroughly in `test/e2e/specs/wp-my-home-support-search-spec.js`. Maybe leave a basic "Can perform a search" and "Can remove the search" to smoke test that it's still working correctly within the popover?

File: test/e2e/specs/wp-my-home-support-search-spec.js (hunk 2 above)
Comment (jeryj) [reply test/e2e/specs/wp-my-home-support-search-spec.js:94]: Moved these FAB tests over to `test/e2e/specs/wp-inline-help-support-search-spec.js ` since it felt more relevant there.

File: test/e2e/lib/components/support-search-component.js [hunk 4]
```diff
@@ -13,9 +13,14 @@ import * as driverHelper from '../driver-helper.js';
 const searchInputSelectors = By.css(
 	'.inline-help__search input[type="search"], .help-search__search input[type="search"]'
 );
-const searchResultsSelectors = By.css(
-	'.inline-help__results-list .inline-help__results-item, .help-search__results-list .help-search__results-item'
+// A little confusing, but this is specific to the default results and _not_ including the errored results (which are the same results).
```
Comment (getdave) [inline test/e2e/lib/components/support-search-component.js:16]: Nice comment 👍

File: test/e2e/lib/components/support-search-component.js [hunk 5]
```diff
@@ -42,6 +47,32 @@ class SupportSearchComponent extends AsyncBaseContainer {
...
+		return driverHelper.waitTillNotPresent( this.driver, defaultResultsSelectors );
+	}
+
+	async getErrorResults() {
+		return await this.driver.findElements( errorResultsSelectors );
+	}
+
+	async getErrorResultsCount() {
+		const results = await this.getErrorResults();
+		return results.length;
```
Comment (getdave) [inline test/e2e/lib/components/support-search-component.js:69]: Worth checking whether `.findElements` _always_ returns an object that has a length. If not you need a fallback.

File: test/e2e/specs/wp-inline-help-support-search-spec.js [hunk 6]
```diff
@@ -28,49 +29,56 @@ before( async function () {
...
+			const sidebarComponent = await SidebarComponent.Expect( driver );
 
-			const isPopoverVisible = await inlineHelpPopoverComponent.isPopoverVisible();
-			assert.equal( isPopoverVisible, true, 'Popover was not opened correctly.' );
-		} );
+			// The "inline help" FAB should not appear on the My Home
+			// because there is already a support search "Card" on that
+			// page. Therefore we select the "Themes" page for our tests
+			// and then we will switch away to the Homepage to verify
+			// that the Inline Help is not shown
```
Comment (getdave) [inline test/e2e/specs/wp-inline-help-support-search-spec.js:61]: Given the restructure of the test order I reckon this comment needs a tweak.

File: test/e2e/specs/wp-inline-help-support-search-spec.js [hunk 7]
```diff
@@ -81,19 +89,19 @@ describe( `[${ host }] Inline Help support search: (${ screenSize }) @parallel`,
 			assert.equal(
 				resultsCount <= 5,
 				true,
-				`Too many search results displayed. Should be less than or equal to 5 (was ${ resultsCount }).`
+				`Too many search results displayed. Should be less than or equal to 5.`
```
Comment (getdave) [inline test/e2e/specs/wp-inline-help-support-search-spec.js:92]: Minor: I wonder whether the int values for the number of results should be stored on the `SupportSearch` helper component as static properties rather than sprinkling across the tests?

File: test/e2e/specs/wp-inline-help-support-search-spec.js [hunk 8]
```diff
@@ -129,5 +137,11 @@ describe( `[${ host }] Inline Help support search: (${ screenSize }) @parallel`,
 				'A search was unexpectedly performed for an empty search query.'
 			);
 		} );
+
+		step( 'Close Inline Help popover', async function () {
```
Comment (getdave) [inline test/e2e/specs/wp-inline-help-support-search-spec.js:141]: I like this addition.

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 9]
```diff
@@ -38,26 +37,10 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
-		assert.equal(
-			isToggleVisible,
-			true,
-			'Inline Help support search was not displayed on a non My Home page.'
-		);
-	} );
+		await loginFlow.loginAndSelectMySite();
 
-	step( 'Select the My Home page', async function () {
+		// The "/home" route is the only one this support search
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:42]: 👍

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 10]
```diff
@@ -78,48 +61,59 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
-			'Inline Help support search was not correctly hidden on Home page.'
-		);
-	} );
-
 	step( 'Displays contextual search results by default', async function () {
 		supportSearchComponent = await SupportSearchComponent.Expect( driver );
-		const resultsCount = await supportSearchComponent.getSearchResultsCount();
-		assert.equal( resultsCount, 5, 'There are no contextual results displayed' );
+		const resultsCount = await supportSearchComponent.getDefaultResultsCount();
+		assert.equal( resultsCount, 5, 'The default results are displayed' );
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:67]: > 'The default results are displayed

This is going to be the error message if the assertion fails. I wonder whether it reads enough like an error.

Ideally, it would describe _what went wrong_ rather than what happened. 

Small detail but will make debugging these tests much easier if/when they break (which is why I wrote them in the first place) 😄

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 11]
```diff
@@ -78,48 +61,59 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
+		const adminResults = await supportSearchComponent.waitForAdminResultsNotToBePresent();
 
-		assert.equal( resultsCount, 5, 'There are no contextual results displayed' );
+		const defaultResultsCount = await supportSearchComponent.getDefaultResultsCount();
+
+		assert.equal( searchResults, true, 'The API Search Results are not displayed.' );
+
+		assert.equal( adminResults, true, 'The Admin Results are not displayed.' );
+
+		assert.equal( defaultResultsCount, 5, 'The default results displayed' );
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:116]: Minor: should we be consistent with `Results` vs `results`?

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 12]
```diff
@@ -128,19 +122,34 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
+			const resultsCount = await supportSearchComponent.getErrorResultsCount();
 
 			const hasNoResultsMessage = await supportSearchComponent.hasNoResultsMessage();
 
+			assert.equal( searchResults, true, 'The API Search Results are not displayed.' );
+
 			assert.equal( hasNoResultsMessage, true, 'The "No results" message was not displayed.' );
 
-			assert.equal( resultsCount, 5, 'There are no contextual results displayed.' );
+			assert.equal( resultsCount, 5, 'The default error results are displayed.' );
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:136]: Again I'm not sure the error message describes what went _wrong_ well enough unless you have a full context and understanding of how the support search works (like you do).

File: test/e2e/specs/wp-my-home-support-search-spec.js [hunk 13]
```diff
@@ -128,19 +122,34 @@ describe( `[${ host }] My Home "Get help" support search card: (${ screenSize })
...
+	step( 'Clearing search resets to default state', async function () {
 		await supportSearchComponent.clearSearchField();
 
+		const errorResults = await supportSearchComponent.waitForErrorResultsNotToBePresent();
+
+		const resultsCount = await supportSearchComponent.getDefaultResultsCount();
+
+		assert.equal( errorResults, true, 'The default error results are not displayed.' );
+
+		assert.equal( resultsCount, 5, 'The default results are displayed' );
```
Comment (getdave) [inline test/e2e/specs/wp-my-home-support-search-spec.js:149]: Same with error message not telling me what went _wrong_.

File: client/blocks/inline-help/inline-help-search-results.jsx [hunk 14]
```diff
@@ -186,6 +186,10 @@ function HelpSearchResults( {
 		} );
 	};
 
+	const resultsLabel = hasAPIResults
```
Comment (getdave) [inline client/blocks/inline-help/inline-help-search-results.jsx:189]: I'd say this should be broken out into it's own PR. However, given the flakiness of the e2e tests at the moment I'm loath to request it.

Agree we try and roll it in with this PR 🤞

**Review by getdave — APPROVED** [review]
Great work jumping in and upgrading these tests.

Usually, I'd say break out the change to the actual implementation and keep e2e test PRs as purely changes to e2e tests. However, given the test failure rate on CI at the moment I'd rather just roll it in this PR this one time 😄 !

I left some minor changes and optional improvements but generally this looks good.

## Testing Results

I tested the `aria-label` changes on the Live link and it all worked as described and it seems logical. I was wondering whether this would be better as a true HTML heading (as the rule of thumb is prefer plain old HTML where possible). But as our resident a11y expert I will defer to your judgement here.

I also manually ran the tests on local and verified they pass. CI also seems to pass so I think we're 👍 

✅ `./node_modules/.bin/mocha specs/wp-inline-help-support-search-spec.js`
✅ `./node_modules/.bin/mocha specs/wp-my-home-support-search-spec.js`

🚢

File: test/e2e/lib/components/support-search-component.js (hunk 5 above)
Comment (jeryj) [reply test/e2e/lib/components/support-search-component.js:69]: From the [Selenium docs](https://www.selenium.dev/documentation/en/getting_started_with_webdriver/locating_elements/), it looks like it will return an empty array (they use "list", which I took to mean "array"):

> findElements(By). This method returns a collection of web elements. If only one element is found, it will still return a collection (of one element). If no element matches the locator, an empty list will be returned

File: test/e2e/specs/wp-my-home-support-search-spec.js (hunk 10 above)
Comment (jeryj) [reply test/e2e/specs/wp-my-home-support-search-spec.js:67]: > Ideally, it would describe what went wrong rather than what happened.

I hadn't considered the error messages from that perspective. That's really helpful to keep in mind. I had changed it because I thought, "This assertion doesn't determine that no contextual results are displayed, I should change it to describe what it does determine." 😅

File: test/e2e/specs/wp-inline-help-support-search-spec.js (hunk 7 above)
Comment (jeryj) [reply test/e2e/specs/wp-inline-help-support-search-spec.js:92]: I'll leave it off this time, but that sounds like a good thing for me to keep in mind for my next round of e2e tests!

File: client/blocks/inline-help/inline-help-search-results.jsx (hunk 14 above)
Comment (jeryj) [reply client/blocks/inline-help/inline-help-search-results.jsx:189]: I tried to leave it off, but it allowed me to truly isolate between the default results list and the errored results lists.

File: test/e2e/specs/wp-my-home-support-search-spec.js (hunk 11 above)
Comment (jeryj) [reply test/e2e/specs/wp-my-home-support-search-spec.js:116]: I went through and tried to make the naming of everything more consistent.

File: test/e2e/specs/wp-my-home-support-search-spec.js (hunk 12 above)
Comment (jeryj) [reply test/e2e/specs/wp-my-home-support-search-spec.js:136]: All assertions are now written with the perspective of, "How would this be most helpful if you encounter the error."

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/4244225

Hi @jeryj, could you please edit the description of this PR and add a screenshot for our translators? Ideally it'd include this string: `Helpful resources for this section`

Thank you in advance!

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/4244225).
